File: Services/YouTubeAPIError.swift
import Foundation

/// YouTube API 関連のエラー。ユーザー向けに日本語メッセージを返す。
enum YouTubeAPIError: LocalizedError, Equatable {
    case invalidURL
    case invalidChannelURL
    case invalidVideoURL
    case videoNotFound
    case channelNotFound
    case uploadsPlaylistNotFound
    case apiKeyMissing
    case quotaExceeded
    case networkError
    case decodingError
    case unknown

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return String(localized: "error.invalidURL")
        case .invalidChannelURL:
            return String(localized: "error.invalidChannelURL")
        case .invalidVideoURL:
            return String(localized: "error.invalidVideoURL")
        case .videoNotFound:
            return String(localized: "error.videoNotFound")
        case .channelNotFound:
            return String(localized: "error.channelNotFound")
        case .uploadsPlaylistNotFound:
            return String(localized: "error.uploadsPlaylistNotFound")
        case .apiKeyMissing:
            return String(localized: "error.apiKeyMissing")
        case .quotaExceeded:
            return String(localized: "error.quotaExceeded")
        case .networkError:
            return String(localized: "error.network")
        case .decodingError:
            return String(localized: "error.decoding")
        case .unknown:
            return String(localized: "error.unknown")
        }
    }
}

File: android/app/src/main/java/com/deskflowlabs/channeltimelineviewer/network/YouTubeApiError.kt
package com.deskflowlabs.channeltimelineviewer.network

import androidx.annotation.StringRes
import com.deskflowlabs.channeltimelineviewer.R

/**
 * YouTube API 関連のエラー。表示文言は iOS 版と同じ翻訳（`Localization/strings.json`）を使う。
 */
enum class YouTubeApiError(@StringRes val messageRes: Int) {
    InvalidUrl(R.string.error_invalidurl),
    InvalidChannelUrl(R.string.error_invalidchannelurl),
    InvalidVideoUrl(R.string.error_invalidvideourl),
    VideoNotFound(R.string.error_videonotfound),
    ChannelNotFound(R.string.error_channelnotfound),
    UploadsPlaylistNotFound(R.string.error_uploadsplaylistnotfound),
    ApiKeyMissing(R.string.error_apikeymissing),
    QuotaExceeded(R.string.error_quotaexceeded),
    NetworkError(R.string.error_network),
    DecodingError(R.string.error_decoding),
    Unknown(R.string.error_unknown),
}

/** 画面に出すエラーを運ぶ例外。メッセージは表示側で文字列リソースから引く。 */
class YouTubeApiException(val error: YouTubeApiError) : Exception(error.name)
